alicloud/data_source_alicloud_gpdb_instances.go

```diff
@@ -0,0 +1,219 @@
+						"description": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"region_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"availability_zone": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"creation_time": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"status": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"engine": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"engine_version": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"instance_class": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"instance_group_count": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"instance_network_type": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"charge_type": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceAlicloudGpdbInstancesRead(d *schema.ResourceData, meta interface{}) error {
+	// name regex
+	var nameRegex *regexp.Regexp
+	if v, ok := d.GetOk("name_regex"); ok {
+		if r, err := regexp.Compile(v.(string)); err == nil {
+			nameRegex = r
+		} else {
+			return WrapError(err)
+		}
+	}
+	// availability zone
+	var availabilityZone string
+	if v, ok := d.GetOk("availability_zone"); ok {
+		availabilityZone = strings.ToLower(v.(string))
+	}
+	// ids
+	idsMap := make(map[string]string)
+	if v, ok := d.GetOk("ids"); ok {
+		for _, vv := range v.([]interface{}) {
+			idsMap[Trim(vv.(string))] = Trim(vv.(string))
+		}
+	}
+
+	client := meta.(*connectivity.AliyunClient)
+	gpdbService := GpdbService{client}
+	request := gpdb.CreateDescribeDBInstancesRequest()
+	request.RegionId = client.RegionId
+	request.PageSize = requests.NewInteger(PageSizeLarge)
+	request.PageNumber = requests.NewInteger(1)
+
+	var dbInstances []gpdb.DBInstanceAttribute
+	for {
+		raw, err := client.WithGpdbClient(func(gpdbClient *gpdb.Client) (interface{}, error) {
+			return gpdbClient.DescribeDBInstances(request)
+		})
+		if err != nil {
+			return WrapErrorf(err, DataDefaultErrorMsg, "alicloud_gpdb_instances", request.GetActionName(), AlibabaCloudSdkGoERROR)
+		}
+		response, _ := raw.(*gpdb.DescribeDBInstancesResponse)
+		addDebug(request.GetActionName(), response)
+		if len(response.Items.DBInstance) < 1 {
+			break
+		}
+
+		for _, item := range response.Items.DBInstance {
+			// filter by description regex
+			if nameRegex != nil {
+				if !nameRegex.MatchString(item.DBInstanceDescription) {
+					continue
+				}
+			}
+			// filter by instance id
+			if len(idsMap) > 0 {
+				if _, ok := idsMap[item.DBInstanceId]; !ok {
+					continue
+				}
+			}
+			// filter by availability zone
+			if len(availabilityZone) > 0 && availabilityZone != strings.ToLower(string(item.ZoneId)) {
```

Comment:
I think `len(availabilityZone) > 0` is redundant.

Reply:
len(availabilityZone) > 0 can't remove, because availabilityZone is OPTIONAL, not MUST

Reply:
> I think `len(availabilityZone) > 0` is redundant.

len(availabilityZone) > 0 can't remove, because availabilityZone is OPTIONAL, not MUST

Reply:
"len(availabilityZone) > 0" can not remove, because availabilityZone is optional, not required.